File: runme.m
%% MATLAB Codes for the Image Inpainting Problem
%
% Authors:
% Simone Parisotto          (email: sp751 at cam dot ac dot uk)
% Carola-Bibiane Schoenlieb (email: cbs31 at cam dot ac dot uk)
%      
% Address:
% Cambridge Image Analysis
% Centre for Mathematical Sciences
% Wilberforce Road
% CB3 0WA, Cambridge, United Kingdom
%  
% Date:
% September, 2016
%
% Licence: BSD-3-Clause (https://opensource.org/licenses/BSD-3-Clause)
%
clear
close all
clc

addpath ./lib
addpath ./dataset

%% AMLE (Absolute Minimizing Lipschitz Extension) Inpainting
clear
close all

% create the corrupted image with the mask
cleanfilename = 'amle_clean.png';
maskfilename  = 'amle_mask.png';
[u,mask]      = create_image_and_mask(cleanfilename,maskfilename);
imwrite(u,'./dataset/amle_input.png')

% parameters
lambda        = 10^2; 
tol           = 1e-8;
maxiter       = 40000;
dt            = 0.01;

% inpainting
tic
inpainting_amle(u,mask,lambda,tol,maxiter,dt);
toc

%% Harmonic Inpainting
clear
close all

% create the corrupted image with the mask
cleanfilename = 'harmonic_clean.png';
maskfilename  = 'harmonic_mask.png';
[u,mask]      = create_image_and_mask(cleanfilename,maskfilename);
imwrite(u,'./dataset/harmonic_input.png')

% parameters
lambda        = 10;
tol           = 1e-5;
maxiter       = 500;
dt            = 0.1;

% inpainting
tic
inpainting_harmonic(u,mask,lambda,tol,maxiter,dt);
toc

%% Mumford-Shah Inpainting
clear
close all

% create the corrupted image with the mask
cleanfilename = 'mumford_shah_clean.png';
maskfilename  = 'mumford_shah_mask.png';
[u,mask]      = create_image_and_mask(cleanfilename,maskfilename);
imwrite(u,'./dataset/mumford_shah_input.png')

% parameters
maxiter       = 20; 
tol           = 1e-14;
param.lambda  = 10^9;   % weight on data fidelity (should usually be large).
param.alpha   = 1;      % regularisation parameters \alpha.
param.gamma   = 0.5;    % regularisation parameters \gamma.
param.epsilon = 0.05;   % accuracy of Ambrosio-Tortorelli approximation of the edge set.

% inpainting
tic
inpainting_mumford_shah(u,mask,maxiter,tol,param);
toc

%% Cahn-Hilliard Inpainting
clear
close all

% create the corrupted image with the mask
cleanfilename = 'cahn_hilliard_clean.png';
maskfilename  = 'cahn_hilliard_mask.png';
[u,mask]      = create_image_and_mask(cleanfilename,maskfilename);
imwrite(u,'./dataset/cahn_hilliard_input.png')

% parameters
maxiter       = 4000; 
param.epsilon = [100 1];
param.lambda  = 10;
param.dt      = 1;

% inpainting
tic
inpainting_cahn_hilliard(u,mask,maxiter,param);
toc

%% Transport Inpainting
clear
close all

% create the corrupted image with the mask
cleanfilename = 'transport_clean.png';
maskfilename  = 'transport_mask.png';
[u,mask]      = create_image_and_mask(cleanfilename,maskfilename);
imwrite(u,'./dataset/transport_input.png')

% parameters
tol           = 1e-5;
maxiter       = 50;
dt            = 0.1;
param.M       = 40; % number of steps of the inpainting procedure;
param.N       = 2;  % number of steps of the anisotropic diffusion;
param.eps     = 1e-10;

% inpainting
tic
inpainting_transport(u,mask,maxiter,tol,dt,param);
toc

File: python/runme.py
# PYTHON Codes for the Image Inpainting Problem
#
# Authors:
# Simone Parisotto          (email: sp751 at cam dot ac dot uk)
# Carola-Bibiane Schoenlieb (email: cbs31 at cam dot ac dot uk)
#      
# Address:
# Cambridge Image Analysis
# Centre for Mathematical Sciences
# Wilberforce Road
# CB3 0WA, Cambridge, United Kingdom
#  
# Date:
# October, 2018
#
# Licence: BSD-3-Clause (https://opensource.org/licenses/BSD-3-Clause)
#

from __future__ import division

import sys
sys.path.append("./lib")

from ttictoc import tic,toc

import matplotlib.image as mpimg
import inpainting

### AMLE Inpainting

# create the corrupted image with the mask
cleanfilename = './dataset/amle_clean.png';
maskfilename  = './dataset/amle_mask.png';
input,mask    = inpainting.create_image_and_mask(cleanfilename,maskfilename);
mpimg.imsave("./dataset/amle_input.png", input[:,:,0], cmap="gray")

# parameters
fidelity      = 10^2; 
tol           = 1e-8;
maxiter       = 40000;
dt            = 0.01;

# inpainting
tic()
u = inpainting.amle(input,mask,fidelity,tol,maxiter,dt)
print('Elasped time is:',toc())

### Harmonic Inpainting

# create the corrupted image with the mask
cleanfilename = './dataset/harmonic_clean.png';
maskfilename  = './dataset/harmonic_mask.png';
input,mask    = inpainting.create_image_and_mask(cleanfilename,maskfilename);
mpimg.imsave("./dataset/harmonic_input.png", input)

# parameters
fidelity      = 10;
tol           = 1e-5;
maxiter       = 500;
dt            = 0.1;

# inpainting
tic()
u = inpainting.harmonic(input,mask,fidelity,tol,maxiter,dt)
print('Elasped time is:',toc())

### Mumford-Shah Inpainting

# create the corrupted image with the mask
cleanfilename = './dataset/mumford_shah_clean.png';
maskfilename  = './dataset/mumford_shah_mask.png';
input,mask    = inpainting.create_image_and_mask(cleanfilename,maskfilename);
mpimg.imsave("./dataset/mumford_shah_input.png", input)

# parameters
maxiter   = 20; 
tol       = 1e-14;
fidelity  = 10^9;   # weight on data fidelity (should usually be large).
alpha     = 1;      # regularisation parameters \alpha.
gamma     = 0.5;    # regularisation parameters \gamma.
epsilon   = 0.05;   # accuracy of Ambrosio-Tortorelli approximation of the edge set.

# inpainting
tic()
u = inpainting.mumford_shah(input,mask,maxiter,tol,fidelity,alpha,gamma,epsilon);
print('Elasped time is:',toc())

### Cahn-Hilliard Inpainting

# create the corrupted image with the mask
cleanfilename = './dataset/cahn_hilliard_clean.png';
maskfilename  = './dataset/cahn_hilliard_mask.png';
input,mask    = inpainting.create_image_and_mask(cleanfilename,maskfilename);
mpimg.imsave("./dataset/cahn_hilliard_input.png", input)

# parameters
maxiter  = 4000
fidelity = 10
dt       = 1

# inpainting
tic()
u = inpainting.cahn_hilliard(input,mask,maxiter,fidelity,dt)
print('Elasped time is:',toc())

### Transport Inpainting

# create the corrupted image with the mask
cleanfilename = './dataset/transport_clean.png';
maskfilename  = './dataset/transport_mask.png';
input,mask    = inpainting.create_image_and_mask(cleanfilename,maskfilename);
mpimg.imsave("./dataset/transport_input.png", input)

# parameters
maxiter          = 50;
tol              = 1e-5;
dt               = 0.1;
iter_inpainting  = 40; # number of steps of the inpainting procedure;
iter_anisotropic = 2;  # number of steps of the anisotropic diffusion;
epsilon          = 1e-10;

# inpainting
tic()
u = inpainting.transport(input,mask,maxiter,tol,dt,iter_inpainting,iter_anisotropic,epsilon)
print('Elasped time is:',toc())
